simple/tests/util/file_match_test.py

```diff
@@ -0,0 +1,124 @@
+import unittest
+
+from util.file_match import match
+from util.filesystem import create_store
+
+
+class TestFileMatch(unittest.TestCase):
+
+  def test_match_nested_dir(self):
+
+    store = create_store("temp://")
+    file = store.as_dir().open_file("path/to/foo.csv")
+
+    def yes(pattern: str) -> None:
+      self.assertTrue(match(file, pattern))
+
+    def no(pattern: str) -> None:
+      self.assertFalse(match(file, pattern))
+
+    def err(pattern: str) -> None:
+      with self.assertRaises(ValueError):
+        match(file, pattern)
+
+    # File path: temp://path/to/foo.csv
+
+    # No slashes: match file name
+    yes("foo.csv")
+    no("bar.csv")
+    no("oo.csv")
+    no("oo*.csv")
+    yes("foo*.csv")  # Wildcard can be no characters
+    yes("f*.csv")
+    yes("*.csv")
+    no("*.mcf")
+
+    # Leading double slash: partial match not allowed, must match from root dir
+    no("//foo.csv")
+    no("//to/foo.csv")
+    yes("//path/to/foo.csv")
+    yes("temp://path/to/**/*")
+    no("//*")
+    yes("//**/*")
+    no("//*/foo.csv")  # Single wildcard is a single level of nesting
+    no("//*/*/*/foo.csv")  # Wrong depth with single wildcards
+    yes("//*/*/foo.csv")  # Right depth with single wildcards
+    no("//*/*/oo.csv")  # Right depth, wrong filename
+    yes("//*/*/*oo.csv")  # Wildcard paths, wildcard in filename
+    yes("//**/foo.csv")  # Double wildcard can be multiple levels of dirs
+    no("//**/oo.csv")
+    yes("//**/*/foo.csv")
+    yes("//*/**/foo.csv")
+    yes("//**/*/*/foo.csv")
+    yes("//*/**/*/foo.csv")
+    yes("//**/*/**/*/**/foo.csv")  # gettin silly with it
+    no("//**/*/**/*/**/*/**/foo.csv")
+    no("//*/**/*/*/foo.csv")
+    no("//**/*/*/*/foo.csv")
+
+    # Leading single slash - equivalent to no leading slash
+    yes("/foo.csv")  # File is not in root dir
```

Comment:
Should a single leading slash be an absolute path? I could see users testing locally and wanting to load data from a known location somewhere else on the file system

Reply:
I'm inclined to use leading double slash to match absolute path instead, for consistency with non-system paths e.g. gs://bucket/some/dir. But we can use the single vs double slash distinction for whether to match the full absolute path or the full relative path (relative to the root of the input dir). WDYT?

Reply:
That's fine with me as long as it's clear to the user.  The word root is used in a few places in the comments which was throwing me off. Maybe we can use "Filesystem root" for a leading `//`, and "Relative to the root of the input dir" (like you have worded in your comment above) for a leading `/`?

Reply:
Done, and updated comments that reference "root" too!